spawn-fcgi should be the prefered way compare to fcgiwrap

Gustry opened the pull request:
<!---
Include a few sentences describing the overall goals for this Pull Request.
 
A list of issues is at https://github.com/qgis/QGIS-Documentation/issues.
Add "fix #issuenumber" for each issue the PR fixes. The ticket(s) will be closed automatically.
If your PR doesn't fix entirely the ticket, only add the ticket(s) reference preceded by # character.
-->
Goal:

A few days ago, I quickly followed the documentation with NGINX and I installed `fcgiwrap`. I didn't see the previous note made by @elemoine. The performance is catastrophic :) according to me. It's indeed very slow. So I think we should move `spawn-fcgi` before `fcgiwrap` in the documentation.

fcgiwrap is easy to setup, but very slow.

I also tried to add a systemd file for starting qgis server.

@elemoine I'm clearly not an expert in this area, a review is welcome.

Ticket(s): #
<!---
Indicate whether the fix should be backported to previous release.
Replace the space between square brackets by a `x` to make it checked.
-->
- [x] Backport to LTR documentation is required

<!---
Reviewing is a process done by community members, mostly on a volunteer basis.
We try to keep the overhead as small as possible and appreciate if you help us.
Please read carefully and ensure you comply with our writing guidelines at
https://docs.qgis.org/testing/en/docs/documentation_guidelines/index.html.
Feel free to ask in a comment or the (qgis-community-team mailing list)
[https://lists.osgeo.org/mailman/listinfo/qgis-community-team] if you have troubles with any item.
--->

DelazJ commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
+
+Either in your bash line or in your init file, launch your QGIS Server by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
+
+..code-block:: bash
```
```suggestion
.. code-block:: bash
```

DelazJ commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
+
+QGIS Server should be available and you can watch the live logs with:
+
+.. code-block::bash
```
```suggestion
.. code-block:: bash
```

DelazJ commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
+
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup `multiwatch` to increase the number of process to 2 for instance:
```
```suggestion
As mentioned above, we can setup ``multiwatch`` to increase
the number of process to 2 for instance:
```
If you want it with a particular display

DelazJ commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
+
+   sudo apt install multiwatch
+
+Either in your bash line or in your init file, launch your QGIS Server by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
```suggestion
Either in your bash line or in your init file, launch your QGIS Server
by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
A bit too long

DelazJ commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,85 @@ QGIS Server is now available at http://localhost/qgisserver.
     `multiwatch <https://redmine.lighttpd.net/projects/multiwatch/wiki>`_ tool, which is also
     packaged in Debian.
 
+Of course, we can add an init script (like a ``qgis-server.service`` file
```
Why moving from "you" to "we"?
```suggestion
Of course, you can add an init script (like a ``qgis-server.service`` file
```

DelazJ commented:
@elemoine may I?

Gustry commented:
Maybe @pblottiere can give a review?

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup ``multiwatch`` to increase
+the number of process to 2 for instance:
```
```suggestion
the number of processes to 2 for instance:
```

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+   sudo apt install multiwatch
+
+Either in your bash line or in your init file, launch your QGIS Server
+by adding this line before the path to the :file:`qgis_mapserv.fcgi`:
```
```suggestion
by adding this line before the path to :file:`qgis_mapserv.fcgi`:
```

pblottiere replied:
I think we should find a better wording for this sentence because it seems a little unclear.

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+         fastcgi_param  SCRIPT_FILENAME /usr/lib/cgi-bin/qgis_mapserv.fcgi;
+     }
+
+Finally, restart NGINX and fcgiwrap to take into account the new configuration:
```
```suggestion
Finally, restart NGINX and **fcgiwrap** to take into account the new configuration:
```

pblottiere replied:
(to be consistent with previous chapters)

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
     packaged in Debian.
 
+Of course, you can add an init script (like a ``qgis-server.service`` file
+with Systemd) to start QGIS Server at boot time or whenever you want. To use a init script with systemd,
```
```suggestion
with Systemd) to start QGIS Server at boot time or whenever you want. To use an init script with systemd,
```

elemoine replied:
s/Systemd/systemd to use the same spelling everywhere. And the official is "systemd" I guess.

pblottiere replied:
Yes you're right, I updated the suggestion.

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+^^^^^^^^
+
+Using `fcgiwrap <https://www.nginx.com/resources/wiki/start/topics/examples/fcgiwrap/>`_
+is much easier to setup than `spawn-fcgi` but it's much slower.
```
```suggestion
is much easier to setup than **spawn-fcgi** but it's much slower.
```

pblottiere replied:
for consistency

pblottiere commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+
+   sudo journalctl -f -u qgis-server
+
+As mentioned above, we can setup ``multiwatch`` to increase
```
maybe `**multiwatch**` to be consistent with `fcgiwrap`? Or is it the other way around? 

@DelazJ what should be preferred between **mytool** and ``mytool``?

elemoine replied:
I actually would not mention multiwatch. I've had bad experience with it. And one could use a systemd template file for multi-processing. But again I think this is going too far the QGIS Server documentation.

pblottiere replied:
>  I've had bad experience with it.

+1 to remove multiwatch from the documentation then

pblottiere commented:
> Maybe @pblottiere can give a review?

I just left a few comments.

And thanks @Gustry, I think it's a good move :+1:

elemoine commented on source/docs/user_manual/working_with_ogc/server/getting_started.rst:
```diff
@@ -207,6 +172,87 @@ QGIS Server is now available at http://localhost/qgisserver.
+   ExecStart=spawn-fcgi -s /var/run/qgisserver.socket -U www-data -G www-data -n /usr/lib/cgi-bin/qgis_mapserv.fcgi
+
+   [Install]
+   WantedBy=multi-user.target
```
I personally wouldn't provide a systemd service file in the QGIS documentation. I think this is going too far, with materials that are outside the scope of QGIS.

pblottiere replied:
@elemoine Mmmm, I see you're point. In this case, we could add the systemd service file in the training manual instead.

elemoine replied:
Yes, there are so many ways to set up QGIS Server, and FastCGI applications in general. The QGIS documentation can't cover it all. The current QGIS Server doc covers Apache + mod_fcgid, and NGINX + fcgiwrap and spawn-fcgi. That sounds sufficient to me.

elemoine commented:
I just left a few comments. These are just my suggestions, take them the way you want :)

pblottiere commented:
> fcgiwrap is easy to setup, but very slow.

Otherwise, after some reflection, I'm wondering if `fcgiwarp` should really be mentioned in the documentation. I think it's probably counterproductive to mention a solution which is considered as slow.

@elemoine @Gustry what do you think?

elemoine commented:
> @elemoine @Gustry what do you think?

I'd keep it, as it's so straightforward to set up on Linux.

pblottiere commented:
> I'd keep it, as it's so straightforward to set up on Linux.

I'm still not convinced to mention a solution which should not be used in production :).

Gustry commented:
I have just updated the PR.

I have no idea if we should keep fcgiwrap.

About the service file, I have removed it. I know the QGIS documentation can't (and mustn't) cover the doc of systemd, multiwatch, spawn-fcgi etc But it's not convenient to go from doc to doc and have always something more to read. That's why I included at the beginning a small service file template that we could easily copy/paste. I don't have a proper solution.

DelazJ commented:
Hi all,
Having a look to the current diff, it's more about moving down the fcgiwrap description and raising from note to warning its conditions of usage. This looks like a good compromise for the issue about mentioning or not that system (I have no idea what you are talking about so no opinion myself).
Any objection to merging?

elemoine commented:
Looks good to me.
